File: backend/src/utils/systemConfig.ts
import { Request, Response } from 'express';
import * as console from 'node:console';
import { prisma } from '@/prisma';

export enum SystemConfigKey {
	BASE_URL = 'BASE_URL:string',

	//Git sync
	GIT_URL = 'GIT_URL:string',
	GIT_BRANCH = 'GIT_BRANCH:string',
	SSH_KEY = 'SSH_KEY:string',
	GIT_NAME = 'GIT_NAME:string',
	GIT_EMAIL = 'GIT_EMAIL:string',

}

// Default values for variables
export const defaultVariables = {
	[SystemConfigKey.BASE_URL]: '',

	//Git sync
	[SystemConfigKey.GIT_URL]: '',
	[SystemConfigKey.GIT_BRANCH]: 'main',
	[SystemConfigKey.SSH_KEY]: '',
	[SystemConfigKey.GIT_NAME]: 'Mockoon Control Panel',
	[SystemConfigKey.GIT_EMAIL]: 'noreply@example.com'

};

/**
 * Get a system configuration value from the database with type conversion
 * @param key - Configuration key to retrieve
 * @returns The configuration value with proper type conversion
 */
export const GetSystemConfig = async (key: SystemConfigKey): Promise<string | number | boolean | string[] | undefined> => {
	try {
		const [keyName, keyType] = key.split(':');
		// Try to get from database first
		const variable = await prisma.systemConfig.findUnique({
			where: {
				key: keyName
			}
		});

		if (variable) {
			return convertValue(variable.value, variable.type);
		}

		// If not found in database, check defaults
		const defaultVar = defaultVariables[key as keyof typeof defaultVariables];
		if (defaultVar) {
			return convertValue(defaultVar, keyType);
		}

		return undefined;
	} catch (error) {
		console.error(`Error fetching system config from database: ${error}`);
		return undefined;
	}
};

/**
 * Set a system configuration value in the database with type validation
 * @param key - Configuration key to set
 * @param value - Value to store
 * @returns True if successful, false otherwise
 */
export const SetSystemConfig = async (
	key: SystemConfigKey,
	value: string | number | boolean | string[]
): Promise<boolean> => {
	try {
		const [keyName, keyType] = key.split(':');
		let stringValue = String(value);
		if (keyType === 'array') {
			stringValue = JSON.stringify(value);
		}

		await prisma.systemConfig.upsert({
			where: { key: keyName },
			update: {
				value: stringValue,
				type: keyType
			},
			create: {
				key: keyName,
				value: stringValue,
				type: keyType
			}
		});

		return true;
	} catch (error) {
		console.error(`Error setting system config in database: ${error}`);
		return false;
	}
};

/**
 * Convert string value to appropriate type
 */
const convertValue = (value: string, type: string): string | number | boolean | string[] => {
	switch (type) {
		case 'number':
			return Number(value);
		case 'boolean':
			return value.toLowerCase() === 'true';
		case 'array':
			try {
				return JSON.parse(value);
			} catch (error) {
				console.log('Error parsing array value from database: ', error);
				return [];
			}
		default:
			return value;
	}
};

type ConfigType = {
	key: string;
	value: string;
	description: string;
	type: string;
}

export const GetAllSystemConfigs = async (req: Request, res: Response) => {
	try {
		const configs = await prisma.systemConfig.findMany();
		res.status(200).json(configs);
	} catch (e) {
		console.log('Error fetching configs', e);
		res.status(500).json({ error: 'Internal server error' });
	}
};

export const SetConfigById = async (req: Request, res: Response) => {
	const { id } = req.params;
	const { key, value, description } = req.body as ConfigType;

	try {
		const updatedConfig = await prisma.systemConfig.update({
			where: { id: parseInt(id) },
			data: {
				key,
				value,
				description,
				updatedAt: new Date()
			}
		});
		res.status(200).json(updatedConfig);
	} catch (e) {
		console.log('Error updating config', e);
		res.status(500).json({ error: 'Internal server error' });
	}
};

export const AddConfig = async (req: Request, res: Response) => {
	const { key, value, description, type } = req.body as ConfigType;
	try {
		const newConfig = await prisma.systemConfig.create({
			data: {
				key,
				value,
				description,
				type
			}
		});
		res.status(201).json(newConfig);
	} catch (e) {
		console.log('Error creating config', e);
		res.status(500).json({ error: 'Internal server error' });
	}
};

File: backend_new/src/utils/systemConfig.go
package utils

import (
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"strconv"
	"strings"

	"mockoon-control-panel/backend_new/src/prisma"
)

// System configuration keys
const (
	// Base config
	BaseURL = "BASE_URL:string"

	// Git sync
	GitURL    = "GIT_URL:string"
	GitBranch = "GIT_BRANCH:string"
	SSHKey    = "SSH_KEY:string"
	GitName   = "GIT_NAME:string"
	GitEmail  = "GIT_EMAIL:string"
)

// DefaultVariables contains default values for system configuration
var DefaultVariables = map[string]string{
	BaseURL:   "",
	GitURL:    "",
	GitBranch: "main",
	SSHKey:    "",
	GitName:   "Mockoon Control Panel",
	GitEmail:  "noreply@example.com",
}

// GetSystemConfig retrieves a system configuration value from the database with type conversion
func GetSystemConfig(key string) (interface{}, error) {
	parts := strings.Split(key, ":")
	if len(parts) != 2 {
		return nil, errors.New("invalid key format, expected key:type")
	}

	keyName := parts[0]
	keyType := parts[1]

	// Try to get from database first
	var config prisma.SystemConfig
	result := prisma.DB.Where("key = ?", keyName).First(&config)
	if result.Error == nil {
		// Return with proper type conversion
		return convertValue(config.Value, config.Type)
	}

	// If not found in database, check defaults
	if defaultVal, exists := DefaultVariables[key]; exists {
		return convertValue(defaultVal, keyType)
	}

	return nil, fmt.Errorf("configuration key %s not found", key)
}

// SetSystemConfig sets a system configuration value in the database with type validation
func SetSystemConfig(key string, value interface{}) error {
	parts := strings.Split(key, ":")
	if len(parts) != 2 {
		return errors.New("invalid key format, expected key:type")
	}

	keyName := parts[0]
	keyType := parts[1]

	// Convert value to string based on type
	var stringValue string
	switch keyType {
	case "array":
		bytes, err := json.Marshal(value)
		if err != nil {
			return fmt.Errorf("failed to marshal array value: %w", err)
		}
		stringValue = string(bytes)
	default:
		stringValue = fmt.Sprintf("%v", value)
	}

	// Check if the key exists
	var config prisma.SystemConfig
	result := prisma.DB.Where("key = ?", keyName).First(&config)

	if result.Error == nil {
		// Update existing config
		config.Value = stringValue
		config.Type = keyType
		if err := prisma.DB.Save(&config).Error; err != nil {
			return fmt.Errorf("failed to update system config: %w", err)
		}
	} else {
		// Create new config
		config = prisma.SystemConfig{
			Key:   keyName,
			Value: stringValue,
			Type:  keyType,
		}
		if err := prisma.DB.Create(&config).Error; err != nil {
			return fmt.Errorf("failed to create system config: %w", err)
		}
	}

	return nil
}

// convertValue converts a string value to the appropriate type based on type string
func convertValue(value string, valueType string) (interface{}, error) {
	switch valueType {
	case "number":
		return strconv.ParseFloat(value, 64)
	case "boolean":
		return strings.ToLower(value) == "true", nil
	case "array":
		var result []string
		err := json.Unmarshal([]byte(value), &result)
		if err != nil {
			log.Printf("Error parsing array value: %v", err)
			return []string{}, nil
		}
		return result, nil
	default:
		return value, nil
	}
}

// GetAllSystemConfigs retrieves all system configurations from the database
func GetAllSystemConfigs() ([]prisma.SystemConfig, error) {
	var configs []prisma.SystemConfig
	if err := prisma.DB.Find(&configs).Error; err != nil {
		return nil, fmt.Errorf("failed to fetch system configs: %w", err)
	}
	return configs, nil
}

// SetConfigByID updates a system configuration by its ID
func SetConfigByID(id int, key, value, description string) (*prisma.SystemConfig, error) {
	var config prisma.SystemConfig
	if err := prisma.DB.First(&config, id).Error; err != nil {
		return nil, fmt.Errorf("config with ID %d not found: %w", id, err)
	}

	config.Key = key
	config.Value = value
	config.Description = description

	if err := prisma.DB.Save(&config).Error; err != nil {
		return nil, fmt.Errorf("failed to update config: %w", err)
	}

	return &config, nil
}

// AddConfig creates a new system configuration
func AddConfig(key, value, description, valueType string) (*prisma.SystemConfig, error) {
	config := prisma.SystemConfig{
		Key:         key,
		Value:       value,
		Description: description,
		Type:        valueType,
	}

	if err := prisma.DB.Create(&config).Error; err != nil {
		return nil, fmt.Errorf("failed to create config: %w", err)
	}

	return &config, nil
}
